### PR TITLE
Complete methods of Character about code points in char sequences.

### LINKED ISSUE #5010: java.lang.Character.codePointAt(java.lang.CharSequence,int)int
I feel like this method could be added to Scala.JS easily.

```
[error] Referring to non-existent method static java.lang.Character.codePointAt(java.lang.CharSequence,int)int

```

Scala.JS version 1.16.0

### PR DESCRIPTION
**Pull request by sjrd**
Implement all overloads of `java.lang.Character.`:

* `codePointAt`
* `codePointBefore`
* `codePointCount`
* `offsetByCodePoints`

In the process, unify with the implementations of methods of the same name in `jl.String`.

This completes the set of methods of `jl.Character` that do not need the Unicode database.

Fixes #5010, among other things.

### CODE REVIEW

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringTest.scala [hunk 1]
```diff
@@ -279,9 +283,9 @@ class StringTest {
     assertEquals(s.length - 1, s.offsetByCodePoints(s.length - 1, 0))
     assertEquals(s.length, s.offsetByCodePoints(s.length, 0))
 
-    assertThrows(classOf[IndexOutOfBoundsException], s.offsetByCodePoints(-3, 4))
-    assertThrows(classOf[IndexOutOfBoundsException], s.offsetByCodePoints(6, 18))
-    assertThrows(classOf[IndexOutOfBoundsException], s.offsetByCodePoints(30, 2))
+    assertThrows(classOf[IndexOutOfBoundsException], s.offsetByCodePoints(-3, -4))
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringTest.scala:286]: Why the negative sign on `-4`? Wouldn't it make more sense to test only the `index` is negative case with this one?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringTest.scala [hunk 2]
```diff
@@ -279,9 +283,9 @@ class StringTest {
...
+    assertThrows(classOf[IndexOutOfBoundsException], s.offsetByCodePoints(6, -18))
+    assertThrows(classOf[IndexOutOfBoundsException], s.offsetByCodePoints(30, -2))
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringTest.scala:288]: Why are the negative signs necessary (IIUC they don't harm / change the test, but it feels the change is unnecessary)?

**Review by gzm0 — CHANGES_REQUESTED** [review]

File: javalib/src/main/scala/java/lang/CharSequence.scala [hunk 3]
```diff
@@ -18,3 +18,31 @@ trait CharSequence {
...
 }
+
+private[lang] object CharSequence {
+  /** Wraps an `Array[Char]` as a `CharSequence` to reuse algorithms.
+   *
+   *  `subSequence` has an inefficient implementation. Avoid using this class
+   *  for algorithms that use that method.
+   */
+  @inline
+  private[lang] def ofArray(array: Array[Char]): OfArray = new OfArray(array)
```
[the remark this answers is not in the document]
Comment (sjrd) [reply javalib/src/main/scala/java/lang/CharSequence.scala:29]: That, plus not reaching all of the `Buffer` things for such small methods. 😉

File: javalib/src/main/scala/java/lang/Character.scala [hunk 4]
```diff
@@ -195,6 +195,74 @@ object Character {
...
+  @noinline
+  def codePointAt(a: Array[Char], index: Int, limit: Int): Int = {
+    // implicit null check and bounds check
+    if (!(limit <= a.length && 0 <= index && index < limit))
+      throw new IndexOutOfBoundsException()
+
+    if (index == limit - 1)
+      a(index).toInt // the only case where `limit` makes a difference
+    else
+      codePointAtImpl(CharSequence.ofArray(a), index)
```
[the remark this answers is not in the document]
Comment (sjrd) [reply javalib/src/main/scala/java/lang/Character.scala:215]: Ah yes, that's a good observation.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringTest.scala (hunk 2 above)
Comment (sjrd) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringTest.scala:288]: Oops, I meant to change the ones in `offsetByCodePointsBackward`.

**Review by sjrd — COMMENTED** [review]
I won't have a computer until Monday evening, so I'll update then.

**Review by gzm0 — APPROVED** [review]
